### PR TITLE
first pass fileinput

### PR DESCRIPTION
**Pull request by britt6612**
#### What does this PR do?
This is adding `fileinput` to the design-system
~~There is some work that is currently being done on Grommet to match `fileinput` to designs
https://github.com/grommet/grommet/pull/5183~~ - Merged
#### Where should the reviewer start?
fileinput
#### Should this PR be mentioned in Design System updates?
yes
#### Is this change backwards compatible or is it a breaking change?
backwards compatible

### CODE REVIEW

File: aries-site/src/data/structures/components.js [hunk 1]
```diff
@@ -100,6 +100,26 @@ export const components = [
       background: 'background-front',
     },
   },
+  {
+    name: 'FileInput',
+    category: 'Inputs',
+    description:
+      'A widget which allows the user to select one or more files to upload.',
```
Comment (taysea) [inline aries-site/src/data/structures/components.js:107]:
```suggestion
      'A file upload control that can be used to upload one or more files.',
```
Aligning with Figma thumbnail.

File: aries-site/src/data/structures/components.js [hunk 2]
```diff
@@ -100,6 +100,26 @@ export const components = [
...
+    seoDescription:
+      'A widget which allows the user to select one or more files to upload.',
```
Comment (taysea) [inline aries-site/src/data/structures/components.js:109]:
```suggestion
      'A file upload control that can be used to upload one or more files.',
```

File: aries-site/src/examples/components/fileinput/FileInputSimpleExample.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+
+import { Box, FileInput } from 'grommet';
+
+export const FileInputSimpleExample = () => (
+  <Box align="center" pad="large">
```
Comment (taysea) [inline aries-site/src/examples/components/fileinput/FileInputSimpleExample.js:6]: Can we set a width of `medium` on this box so the file input is a little wider by default? Right now it's quite narrow:
Screen Shot 2021-05-03 at 11 57 07 AM

File: aries-site/src/examples/components/fileinput/FileInputSimpleExample.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    <FileInput
+      messages={{
+        dropPrompt: 'Drag and Drop',
```
Comment (taysea) [inline aries-site/src/examples/components/fileinput/FileInputSimpleExample.js:9]: Designs seem to call for `Drag and drop or`

**Review by taysea — COMMENTED** [review]
I like the idea of show FileInput in a "real" context, but it feels a little complex to have the DataTable as the main portion of the example / took me time to find the FileInput. I'm wondering if the "Multiple Files" example is more similar to the Simple example, then we link off to a Form example where we should FileInput inline in the Form? Similar to what we did on Card Page.

File: aries-site/src/pages/components/fileinput.mdx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+The FileInput can be used with both the `Space` and `Enter` button to achieve the action.
+FileInput should have clear label that describes the type of files that are needed.
+
+## Variants
+
+FileInputs can be used as a single file upload or multiple files.
+
+### FileInput with a single file
+
+<Example code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/fileinput/FileInputSingleExample.js">
```
Comment (taysea) [inline aries-site/src/pages/components/fileinput.mdx:38]: We should add the `docs` prop and link to grommet site docs for FileInput.

File: aries-site/src/pages/components/fileinput.mdx [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  <FileInputSingleExample />
+</Example>
+
+### FileInput with multiple files
+
+<Example code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/fileinput/FileInputMultipeExample.js">
```
Comment (taysea) [inline aries-site/src/pages/components/fileinput.mdx:44]: We should add the `docs` prop and link to grommet site docs for FileInput.

File: aries-site/src/examples/components/fileinput/FileInputSimpleExample.js (hunk 4 above)
Comment (britt6612) [reply aries-site/src/examples/components/fileinput/FileInputSimpleExample.js:9]: Think its just Drag and drop
Screen Shot 2021-05-05 at 8 06 35 AM

File: aries-site/src/themes/aries.js [hunk 7]
```diff
@@ -1,11 +1,28 @@
...
 import { deepMerge } from 'grommet/utils';
+import { Trash } from 'grommet-icons/icons/Trash';
 
 export const aries = deepMerge(hpe, {
   defaultMode: 'dark',
   // To be stripped out once theme changes are made in grommet-theme-hpe
   // keeping file for use as playground for future theme adjusments that need
   // to be quickly tested
+  fileInput: {
+    button:{
```
Comment (taysea) [inline aries-site/src/themes/aries.js:11]: seems like we need to add the font-weight bold to the button

File: aries-site/src/examples/components/fileinput/FileInputSimpleExample.js (hunk 3 above)
Comment (taysea) [reply aries-site/src/examples/components/fileinput/FileInputSimpleExample.js:6]: I think we should still set a fixed width so it doesn't grow and shrink as files are dragged onto it

**Comment by britt6612** [comment]
```
import React, { useState } from ‘react’;
import PropTypes from ‘prop-types’;
import styled from ‘styled-components’;
import {
  Box,
  Button,
  DataTable,
  FileInput,
  Form,
  FormField,
  Layer,
  Header,
  Heading,
  Menu,
  RadioButtonGroup,
  ResponsiveContext,
  Select,
  Text,
  TextArea,
  TextInput,
} from ‘grommet’;
import { Close } from ‘grommet-icons’;
const data = [
  {
    service: ‘Mercury’,
    tenant: ‘Boeing’,
    requester: ‘jane@boeing.com’,
    quoteName: ‘My_Mercury_1’,
    quoteID: ‘QUO_957_59’,
  },
  {
    service: ‘MCC’,
    tenant: ‘Domain’,
    requester: ‘roman@domain.com’,
    quoteName: ‘VMaas for R&D’,
    quoteID: ‘QUO_123_55’,
  },
  {
    service: ‘Mercury’,
    tenant: ‘Nestle’,
    requester: ‘klaus@nestle.com’,
    quoteName: ‘My_Mercury_QT_1’,
    quoteID: ‘QUO_128_34’,
  },
  {
    service: ‘HPE GLC Private’,
    tenant: ‘Boeing’,
    requester: ‘davina@boeing.com’,
    quoteName: ‘Gem_1’,
    quoteID: ‘QUO_489_39’,
  },
  {
    service: ‘Mercury’,
    tenant: ‘Boeing’,
    requester: ‘davina@boeing.com’,
    quoteName: ‘My_Mercury_QT1’,
    quoteID: ‘QUO_397_30’,
  },
  {
    service: ‘MCC’,
    tenant: ‘Boeing’,
    requester: ‘roman@boeing.com’,
    quoteName: ‘VMaaS for R&D’,
    quoteID: ‘QUO_279_27’,
  },
  {
    service: ‘GLMI’,
    tenant: ‘Boeing’,
    requester: ‘klaus@boeing.com’,
    quoteName: ‘My_Mercury_QT_1’,
    quoteID: ‘QUO_280_38’,
  },
  {
    service: ‘Mercury’,
    tenant: ‘Domain’,
    requester: ‘davina@boeing.com’,
    quoteName: ‘My_Mercury_QT_1’,
    quoteID: ‘QUO_280_38’,
  },
  {
    service: ‘Mercury’,
    tenant: ‘Boeing’,
    requester: ‘jane@boeing.com’,
    quoteName: ‘My_Mercury_1’,
    quoteID: ‘QUO_957_59’,
  },
];
const columns = [
  {
    property: ‘tenant’,
    header: ‘Tenant’,
    render: datum => <Text truncate>{datum.tenant}</Text>,
  },
  {
    property: ‘requester’,
    header: ‘Requester’,
    render: datum => <Text truncate>{datum.requester}</Text>,
    size: ‘xsmall’,
  },
  {
    property: ‘quoteName’,
    header: ‘Quote Name’,
    render: datum => <Text truncate>{datum.quoteName}</Text>,
  },
  {
    property: ‘quoteID’,
    header: ‘Quote ID’,
    render: datum => <Text truncate>{datum.quoteID}</Text>,
    align: ‘end’,
  },
];
const States = [‘Colorado’, ‘California’, ‘Texas’, ‘florida’];
const LayerForm = ({ setOpen }) => {
  // eslint-disable-next-line no-unused-vars
  const onSubmit = ({ value, touched }) => {
    // Your submission logic here
    setOpen(false);
  };
  return (
    <Box gap=“medium”>
      <Box>
        <Header align=“start” pad={{ horizontal: ‘xxsmall’ }}>
          <Box gap=“xxsmall”>
            <Heading level={3} margin=“none”>
              New Quote
            </Heading>
            <Text>
              Please fill out the form and select a file to add a new quote
            </Text>
          </Box>
          <Button icon={<Close />} onClick={() => setOpen(false)} />
        </Header>
      </Box>
      <Form
        validate=“blur”
        method=“post”
        onSubmit={({ value, touched }) => onSubmit({ value, touched })}
      >
        <FormField
          label=“Service Name”
          htmlFor=“service-name-input”
          name=“serviceNameInput”
        >
          <TextInput value=“Mercury” id=“service-name” name=“serviceName” />
        </FormField>
        <FormField
          label=“Quote Name”
          htmlFor=“quote-name-input”
          name=“quoteNameInput”
        >
          <TextInput value=“My_Mercury_QT_2" id=“quote-name” name=“quoteName” />
        </FormField>
        <FormField
          label=“Quote ID”
          htmlFor=“quote-id-input”
          name=“quoteIDInput”
        >
          <TextInput value=“QUO_182_$6" id=“quote-id” name=“quoteID” />
        </FormField>
        <FormField label=“Tenant Type” htmlFor=“tenant-type” name=“tenantType”>
          <RadioButtonGroup
            options={[‘Boeing’, ‘Domain’, ‘Nestle’]}
            id=“tenant-name”
            value=“Domain”
            name=“tenantName”
          />
        </FormField>
        <FormField label=“State” htmlFor=“state-select” name=“stateSelect”>
          <Select value=“Colorado” options={States} id=“state” name=“state” />
        </FormField>
        <FormField
          label=“Description (Optional)”
          htmlFor=“description-textArea”
          name=“Description”
        >
          <TextArea id=“Description” name=“Description” />
        </FormField>
        <FormField id=“FileInput” name=“FileInput”>
          <FileInput
            messages={{
              dropPromptMultiple: ‘Drag and Drop’,
              browse: ‘Select Files’,
            }}
            multiple
            id=“fileInput”
            name=“fileInput”
          />
        </FormField>
        <Box align=“start” margin={{ top: ‘medium’, bottom: ‘small’ }}>
          <Button label=“Submit” primary type=“submit” />
        </Box>
      </Form>
    </Box>
  );
};
LayerForm.propTypes = {
  setOpen: PropTypes.func.isRequired,
};
const ActionsMenu = styled(Menu)`
  border: 1px solid
    ${({ theme }) => theme.global.colors.border[theme.dark ? ‘dark’ : ‘light’]};
`;
export const FileInputMultipeExample = () => {
  const size = React.useContext(ResponsiveContext);
  const [open, setOpen] = useState(false);
  const onClose = () => setOpen(undefined);
  return (
    <>
      <Heading level={3} margin=“none”>
        Quotes
      </Heading>
      <Box
        direction=“row”
        fill=“horizontal”
        justify=“end”
        pad={{ vertical: ‘small’ }}
      >
        <ActionsMenu
          label=“Actions”
          items={[
            {
              label: ‘Add New Quote’,
              onClick: () => {
                setOpen(true);
              },
            },
          ]}
        />
      </Box>
      <Box height={{ max: ‘large’ }} overflow=“auto”>
        <DataTable
          data={data}
          columns={[
            { property: ‘service’, header: ‘Service’, pin: size === ‘small’ },
            ...columns,
          ]}
          pin={size === ‘small’}
        />
      </Box>
      {open && (
        <Layer
          position=“right”
          full={size !== ‘small’ ? ‘vertical’ : true}
          modal
          onClickOutside={onClose}
          onEsc={onClose}
        >
          <Box
            fill=“vertical”
            overflow=“auto”
            width={size !== ‘small’ ? ‘medium’ : undefined}
            pad=“medium”
          >
            <LayerForm setOpen={value => setOpen(value)} />
          </Box>
        </Layer>
      )}
    </>
  );
};
```

Code for a more rich example  like on templates on this page examples should be simple

**Review by taysea — APPROVED** [review]

File: aries-site/src/pages/components/fileinput.mdx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  <FileInputSimpleExample />
+</Example>
+
+## Guidance
+
+FileInput is a specialized input that provides the user with a means of selecting a file or files for upload.
+
+### Usage
+
+- User must be able to upload a file by clicking the "Select File" button, drag and dropping over the drop region, or tabbing to the FileInput and pressing Space/Enter.
```
Comment (halocline) [inline aries-site/src/pages/components/fileinput.mdx:22]:
```suggestion
- User must be able to upload a file by clicking the "Select File" button, dragging and dropping a file over the drop region, or tabbing to the FileInput and pressing the Spacebar or Enter key.
```

File: aries-site/src/pages/components/fileinput.mdx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+- When FileInput is in a Form, the file should not be uploaded/saved to a server until the user has submitted the form.
+- Once a file has been added to the input, a Trash button will appear next to the file name to allow removal of that file.
```
Comment (halocline) [inline aries-site/src/pages/components/fileinput.mdx:24]:
```suggestion
- Once a file has been added to the input, a button featuring a Trash icon will appear next to the file name allowing removal of that file.
```

File: aries-site/src/pages/components/fileinput.mdx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+- The "Select File" button should always be visible, allowing the user to add more files/replace a file after they have already selected one.
+
+For additional configurations, such as defining accepted file type, all instrinsic HTML properties are supported. For a comprehensive list of these options, you can read more about [HTML input type="file"](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file).
+
+### Accessibility
+
+When using FileInput within a Form, the FormField wrapping the FileInput should have `htmlFor` and `name` properties that match the `id` and `name` properties placed onto the FileInput. See the [FileInput within a Form example](/templates/forms#required-and-optional-fields) for reference on how to implement.
```
Comment (halocline) [inline aries-site/src/pages/components/fileinput.mdx:31]:
```suggestion
When using FileInput within a Form, the FormField wrapping the FileInput should have `htmlFor` and `name` properties that match the `id` and `name` properties placed onto the FileInput. Reference the [FileInput within a Form example](/templates/forms#required-and-optional-fields) to see how to implement.
```

File: aries-site/src/pages/components/fileinput.mdx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+## Variants
+
+FileInputs can be used as a single file upload or multiple files.
```
Comment (halocline) [inline aries-site/src/pages/components/fileinput.mdx:35]:
```suggestion
FileInputs may be configured for uploading single or multiple files.
```

File: aries-site/src/themes/aries.js [hunk 12]
```diff
@@ -1,11 +1,50 @@
 import { hpe } from 'grommet-theme-hpe';
 import { deepMerge } from 'grommet/utils';
+import { Trash } from 'grommet-icons/icons/Trash';
 
 export const aries = deepMerge(hpe, {
   defaultMode: 'dark',
   // To be stripped out once theme changes are made in grommet-theme-hpe
   // keeping file for use as playground for future theme adjusments that need
   // to be quickly tested
+  fileInput: {
```
Comment (halocline) [inline aries-site/src/themes/aries.js:10]: Is this work to add to HPE theme represented in an issue?

File: aries-site/src/pages/components/fileinput.mdx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+  docs="https://v2.grommet.io/fileinput?theme=hpe"
+  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/fileinput/FileInputSimpleExample.js"
+  figma="https://www.figma.com/file/8HwAiiSDloOTjUxqvCTypq/HPE-FileInput-Component"
+>
+  <FileInputSimpleExample />
+</Example>
+
+## Guidance
+
+FileInput is a specialized input that provides the user with a means of selecting a file or files for upload.
```
Comment (halocline) [inline aries-site/src/pages/components/fileinput.mdx:18]:
```suggestion
FileInput is a specialized input providing users a means to select a file or files for upload.
```

**Review by halocline — COMMENTED** [review]
Primarily copy suggestions. One question regarding whether an issue exists to implement in the theme.

**Comment by halocline** [comment]
@taysea would you mind reviewing my proposed changes?

File: aries-site/src/themes/aries.js (hunk 12 above)
Comment (taysea) [reply aries-site/src/themes/aries.js:10]: I will file a PR on the theme with this work.

**Comment by taysea** [comment]
> @taysea would you mind reviewing my proposed changes?

Changes looked good. I've incorporated them. Filing a PR on the theme now.

File: aries-site/src/themes/aries.js (hunk 12 above)
Comment (taysea) [reply aries-site/src/themes/aries.js:10]: PR is file here on theme: https://github.com/grommet/grommet-theme-hpe/pull/191

Comment (taysea) [reply aries-site/src/themes/aries.js]: This theme worked has been merged.

File: aries-site/src/themes/aries.js [hunk 14]
```diff
@@ -1,5 +1,6 @@
 import { hpe } from 'grommet-theme-hpe';
 import { deepMerge } from 'grommet/utils';
+import { Trash } from 'grommet-icons/icons/Trash';
```
Comment (halocline) [inline aries-site/src/themes/aries.js:3]:
```suggestion
```

**Review by halocline — APPROVED** [review]

**Pull request merged**